From e0522b41ab13e734fe7aad2ff699c61cbdd57729 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:23:58 -0400
Subject: [tools] perf/build: Speed up auto-detection

Message-id: <1408462094-14194-70-git-send-email-jolsa@redhat.com>
Patchwork-id: 88052
O-Subject: [PATCH RHEL7.1 BZ1131394 069/325] tools/perf/build: Speed up auto-detection
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit f1138ec66e839ce90c58af0c264db33271d73466
Author: Ingo Molnar <mingo@kernel.org>
Date: Wed Oct 2 09:54:43 2013 +0200

description
===========
The detection of certain rarely detected features can be delayed
to when they are actually needed.

So speed up the common case of auto-detection by pre-building only
a core set of features and populating only their feature-flags.

[ Features not listed in CORE_FEATURES need to built explicitly
  via the feature_check() function. ]

(Also order the feature names alphabetically, while at it.)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/config/Makefile b/tools/perf/config/Makefile
index cbd7cdc..581a942 100644
--- a/tools/perf/config/Makefile
+++ b/tools/perf/config/Makefile
@@ -105,30 +105,36 @@ endef
 $(info )
 $(info Auto-detecting system features:)
 
-FEATURE_TESTS =    \
- hello    \
- stackprotector-all  \
- stackprotector   \
- volatile-register-var  \
+#
+# Note that this is not a complete list of all feature tests, just
+# those that are typically built on a fully configured system.
+#
+# [ Feature tests not mentioned here have to be built explicitly in
+#   the rule that uses them - an example for that is the 'bionic'
+#   feature check. ]
+#
+CORE_FEATURE_TESTS =   \
+ backtrace   \
+ dwarf    \
  fortify-source   \
- bionic    \
- libelf    \
  glibc    \
- dwarf    \
- libelf-mmap   \
- libelf-getphdrnum  \
- libunwind   \
- libaudit   \
- libslang   \
  gtk2    \
  gtk2-infobar   \
+ libaudit   \
+ libbfd    \
+ libelf    \
+ libelf-getphdrnum  \
+ libelf-mmap   \
+ libnuma    \
  libperl    \
  libpython   \
  libpython-version  \
- libbfd    \
+ libslang   \
+ libunwind   \
  on-exit    \
- backtrace   \
- libnuma
+ stackprotector   \
+ stackprotector-all  \
+ volatile-register-var
 
 #
 # Special fast-path for the 'all features are available' case:
@@ -136,10 +142,13 @@ FEATURE_TESTS =    \
 $(call feature_check,all)
 
 ifeq ($(feature-all), 1)
-  $(foreach feat,$(FEATURE_TESTS),$(call feature_set,$(feat)))
+  #
+  # test-all.c passed - just set all the core feature flags to 1:
+  #
+  $(foreach feat,$(CORE_FEATURE_TESTS),$(call feature_set,$(feat)))
 else
-  $(shell $(MAKE) -i -j -C config/feature-checks >/dev/null 2>&1)
-  $(foreach feat,$(FEATURE_TESTS),$(call feature_check,$(feat)))
+  $(shell $(MAKE) -i -j -C config/feature-checks $(CORE_FEATURE_TESTS) >/dev/null 2>&1)
+  $(foreach feat,$(CORE_FEATURE_TESTS),$(call feature_check,$(feat)))
 endif
 
 feature_print = $(eval $(feature_print_code))
@@ -156,7 +165,7 @@ define feature_print_code
   $(info $(MSG))
 endef
 
-$(foreach feat,$(FEATURE_TESTS) DUMMY,$(call feature_print,$(feat)))
+$(foreach feat,$(CORE_FEATURE_TESTS) DUMMY,$(call feature_print,$(feat)))
 
 # newline at the end of the feature printouts:
 $(info )
diff --git a/tools/perf/config/feature-checks/Makefile b/tools/perf/config/feature-checks/Makefile
index 4b855e0..d4c55ac 100644
--- a/tools/perf/config/feature-checks/Makefile
+++ b/tools/perf/config/feature-checks/Makefile
@@ -1,28 +1,29 @@
 
 FILES=     \
- test-hello   \
- test-stackprotector-all  \
- test-stackprotector  \
- test-volatile-register-var \
- test-fortify-source  \
+ test-all   \
+ test-backtrace   \
  test-bionic   \
- test-libelf   \
- test-glibc   \
  test-dwarf   \
- test-libelf-mmap  \
- test-libelf-getphdrnum  \
- test-libunwind   \
- test-libaudit   \
- test-libslang   \
+ test-fortify-source  \
+ test-glibc   \
  test-gtk2   \
  test-gtk2-infobar  \
+ test-hello   \
+ test-libaudit   \
+ test-libbfd   \
+ test-libelf   \
+ test-libelf-getphdrnum  \
+ test-libelf-mmap  \
+ test-libnuma   \
  test-libperl   \
  test-libpython   \
  test-libpython-version  \
- test-libbfd   \
+ test-libslang   \
+ test-libunwind   \
  test-on-exit   \
- test-backtrace   \
- test-libnuma
+ test-stackprotector-all  \
+ test-stackprotector  \
+ test-volatile-register-var
 
 CC := $(CC) -MD
 
-- 
1.7.1